lib/spack/llnl/util/filesystem.py

```diff
@@ -974,6 +974,43 @@ def remove_linked_tree(path):
             shutil.rmtree(path, True)
 
 
+@contextmanager
+def safe_remove(*files_or_dirs):
+    """Context manager to remove the files passed as input, but restore
+    them in case any exception is raised in the context block.
+
+    Filenames clash from single files in different directories are currently
+    not handled.
+
+    Args:
+        *files_or_dirs: glob expressions for files or directories
+            to be removed
+
+    Returns:
+        Dictionary that maps deleted files to their temporary copy
+        within the context block.
+    """
+    # Find all the files or directories that match
+    glob_matches = [glob.glob(x) for x in files_or_dirs]
+
+    # Copy them in a temporary location
+    removed, dst_root = {}, tempfile.mkdtemp()
+
+    try:
+        for file_or_dir in itertools.chain(*glob_matches):
```

Comment:
It occurs to me that if you have a mixture of files or directories where some file is contained in one of the directories to remove, there could be an issue:

`[x/, x/y]`

it looks like in this case `x/` would be removed before `x/y` and there would be an error when you got to `x/y`.

I think requiring that the user provide absolute paths would simplify this issue (though you may have to sort and check).

Reply:
I don't think we need to require absolute path, but this is definitely a case worth testing that is failing in the current implementation. Thanks for spotting it!

Reply:
I improved the context manager to handle this case and added this unit test:

https://github.com/spack/spack/blob/9063f3b5b80a5da9afde082a2497663d64a2bae5/lib/spack/spack/test/llnl/util/filesystem.py#L530-L533